### DISCUSSION TITLE
Sound effect

### DISCUSSION THREAD

**Discussion by SyXavier** [schreibfaul1/ESP32-audioI2S#90] [General]
Can a digital audio equalizer code be added to the library to improve the sound quality?

**Comment by schreibfaul1** [comment]
Hi SyXavier,
I've added the IIR_filters () routine to the library. This makes it possible to build a high, low or bandpass. The necessary coefficients can be calculated online: https://www.easycalculation.com/physics/electromagnetism/biquad-calculator.php

**Reply by SyXavier** [reply]
Thank you for taking my advice.

**Comment by ZnakZorro** [comment]
The IIR filter works great. THX.
Is it possible to change IIR filter parameters depending on whether the music or speech is playing?

**Reply by ZnakZorro** [reply]
Or depending on which station is currently playing

**Reply by schreibfaul1** [reply]
The effect depends on the sample rate. The coefficients must be recalculated each time the sample rate is changed. https://www.earlevel.com/main/2012/11/26/biquad-c-source-code/

**Reply by ZnakZorro** [reply]
Thank you for a very valuable link.

**Comment by schreibfaul1** [comment]
There is now a built-in high or low pass filter for each channel.
Example: audio.setTone (0, 5000, 1, 1000); // left lowpass 0...5KHz, right highpass 1KHz ... (sampleRate/2) KHz
The filter coefficients are recalculated each time the sample rate is changed.

Audio.setTone (0, 0, 0, 0); // the filters are turned off

**Comment by hausnerd** [comment]
Do you think about 3 Band equalizer. Where I can ajust the gain, frequenz and Q. So I can ajust speakers that are not very good.
In the earlevel calculator is called peak filter. Three of them are then may usefull. Different filters for the right and left Channel are not so
often usefull.

**Reply by SyXavier** [reply]
I think 3 Band equalizer works very well, thanks.

**Reply by schreibfaul1** [reply]
A lot is possible: with two identical filters in a row, the filter curve becomes steeper. As an equalizer you need (three or more) filters (as bandpass) in parallel? But I'm not very familiar with digital filters.

**Reply by hausnerd** [reply]
In some HiFi oder Professional Speakers they use Parametric Equilizer I described before, to ajust the nonlinear frequenzy characteristic vom Cheap Speakers. Its the same in Alexa Speakers or Google Home. They also use this to make the Sound very good. That the same reason what I think to use Esp32 to make a little soundbox that sounds like bigger.

**Comment by schreibfaul1** [comment]
Hi,
I've removed the old routine with the separate channels. There is now a lowshelf, a combined bandpass / notch and a highshelf. The whole thing is called with: 
setTone (int8_t gainLowPass, int8_t gainBandPass, int8_t gainHighPass);
possible values ​​are -40 ... 6 [dB]. The first tests were successful.

**Reply by hausnerd** [reply]
Oh, it looks very nice I have look at your code. When user can set the frequenzy of this three filters and the Q of the Peak then it is perfect. Because some speakers make different fault in sound it may be usefull to ajust it in frequenz and the wide of the peak. 
Its now at konstant in your code. Its easy to change for me. But I think not for all.
The Hometheater Amplifiers do that also (sometimes automatic with a testmikrofon). I test your code next days.

**Reply by hausnerd** [reply]
Today I test your new setTone.. It work very well, thanks for the very fast response.

**Reply by schreibfaul1** [reply]
Thank you! The filters are modularly in a chain, all are the same So it shouldn't be a problem to add more or change their properties. I determined the frequencies experimentally. Even 18KHz still works fine, I can't hear that, but I can see it on the Oszi.
